packages/utils/src/cryptography/merkle_proof.cairo

```diff
@@ -0,0 +1,149 @@
+}
+
+/// Returns the rebuilt hash obtained by traversing a Merkle tree up
+/// from `leaf` using `proof`. A `proof` is valid if and only if the rebuilt
+/// hash matches the root of the tree. When processing the proof, the pairs
+/// of leaves & pre-images are assumed to be sorted.
+pub fn process_proof<impl Hasher: CommutativeHasher>(
+    proof: Span<felt252>, leaf: felt252
+) -> felt252 {
+    let mut computed_hash = leaf;
+    for hash in proof {
+        computed_hash = Hasher::commutative_hash(computed_hash, *hash);
+    };
+    computed_hash
+}
+
+/// Returns True if the `leaves` can be simultaneously proven to be a part of a Merkle tree defined
+/// by `root`, according to `proof` and `proof_flags` as described in `process_multi_proof`.
+///
+/// CAUTION: Not all Merkle trees admit multiproofs. See `process_multi_proof` for details.
+///
+/// NOTE: Consider the case where `root == proof[0] && leaves.len() == 0` as it will return `True`.
+/// The `leaves` must be validated independently. See `process_multi_proof`.
+pub fn verify_multi_proof<impl Hasher: CommutativeHasher>(
+    proof: Span<felt252>, proof_flags: Span<bool>, root: felt252, leaves: Span<felt252>
+) -> bool {
+    process_multi_proof::<Hasher>(proof, proof_flags, leaves) == root
+}
+
+/// Returns the root of a tree reconstructed from `leaves` and sibling nodes in `proof`. The
+/// reconstruction proceeds by incrementally reconstructing all inner nodes by combining a
+/// leaf/inner node with either another leaf/inner node or a proof sibling node, depending on
+/// whether each `proof_flags` item is true or false respectively.
+///
+/// CAUTION: Not all Merkle trees admit multiproofs. To use multiproofs, it is sufficient to ensure
+/// that: 1) the tree is complete (but not necessarily perfect), 2) the leaves to be proven are in
+/// the opposite order they are in the tree (i.e., as seen from right to left starting at the
+/// deepest layer and continuing at the next layer).
+///
+/// NOTE: The _empty set_ (i.e. the case where `proof.len() == 1 && leaves.len() == 0`) is
+/// considered a no-op, and therefore a valid multiproof (i.e. it returns `proof.at(0)`). Consider
+/// disallowing this case if you're not validating the leaves elsewhere.
+pub fn process_multi_proof<impl Hasher: CommutativeHasher>(
+    proof: Span<felt252>, proof_flags: Span<bool>, leaves: Span<felt252>
+) -> felt252 {
+    // This function rebuilds the root hash by traversing the tree up from the leaves. The root is
+    // rebuilt by consuming and producing values on a queue. The queue starts with the `leaves`
+    // span, then goes onto the `hashes` span. At the end of the process, the last hash in the
+    // `hashes` span should contain the root of the Merkle tree.
+    let leaves_len = leaves.len();
+    let proof_flags_len = proof_flags.len();
+
+    // Check proof validity.
+    if (leaves_len + proof.len() != proof_flags_len + 1) {
+        panic!("MerkleProof: invalid multi proof");
+    }
```

Comment:
It looks like a perfect case to use `assert_eq`

Reply:
assert_eq only works on test.